### PR TITLE
ref(catalog): improve performance of ListAllowedEndpointsForService()

### PR DESCRIPTION
**Pull request by allenlsy**
Signed-off-by: Allen Leigh <allenlsy@gmail.com>

**Description**:

Resolves issue #3190 

* Use map to iterate over endpoint list, in order to improve performace
* Rename `ListAllowedEndpointsForService()` to `ListEndpointsForServiceIdentity()`
* Rename `destSvcAccount` variable to `destinationServiceIdentity`

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | ✅ |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | ✅ |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

No

1. Is this a breaking change?

No

### CODE REVIEW

File: pkg/catalog/endpoint.go [hunk 1]
```diff
@@ -38,17 +38,21 @@ func (mc *MeshCatalog) GetResolvableServiceEndpoints(svc service.MeshService) ([
...
 // for the given downstream identity
 // Note: ServiceIdentity must be in the format "name.namespace" [https://github.com/openservicemesh/osm/issues/3188]
-func (mc *MeshCatalog) ListAllowedEndpointsForService(downstreamIdentity identity.ServiceIdentity, upstreamSvc service.MeshService) ([]endpoint.Endpoint, error) {
+func (mc *MeshCatalog) ListEndpointsForServiceIdentity(downstreamIdentity identity.ServiceIdentity, upstreamSvc service.MeshService) ([]endpoint.Endpoint, error) {
 	outboundEndpoints, err := mc.listEndpointsForService(upstreamSvc)
 	if err != nil {
 		log.Error().Err(err).Msgf("Error looking up endpoints for upstream service %s", upstreamSvc)
 		return nil, err
 	}
+	outboundEndpointsIPSet := make(map[string]endpoint.Endpoint)
```
Comment (shashankram) [inline pkg/catalog/endpoint.go:50]: Use a mapset here instead?
You could use the return value from `set.Add()` to determine if an add was successful, in which case the endpoint can be appended to the final slice.

File: pkg/catalog/endpoint.go [hunk 2]
```diff
@@ -38,17 +38,21 @@ func (mc *MeshCatalog) GetResolvableServiceEndpoints(svc service.MeshService) ([
...
+	for _, ep := range outboundEndpoints {
+		outboundEndpointsIPSet[ep.IP.String()] = ep
```
Comment (shashankram) [inline pkg/catalog/endpoint.go:52]: Isn't this logic flawed, where an endpoint's port is disregarded? Multiple endpoints with the same IP but different ports constitute multiple endpoints, but with this logic they all map to the same key.

**Review by shashankram — CHANGES_REQUESTED** [review]

File: pkg/catalog/endpoint.go (hunk 2 above)
Comment (allenlsy) [reply pkg/catalog/endpoint.go:52]: The current logic is also only checking the IP address: https://github.com/openservicemesh/osm/issues/3190#L63

But I agree that port should be considered during the allowing process.

Comment (shashankram) [reply pkg/catalog/endpoint.go:52]: It seems like the current logic is matching on pod IP only to filer out the endpoints corresponding to the pod. But this logic is disregarding endpoints that have the same port completely.

Comment (allenlsy) [reply pkg/catalog/endpoint.go:52]: I have another question, why do we only compare IP addresses in the first place?

Assume 3 services, down-1, up-1, up-2. 

* down-1 ---denied--> up-1, which has endpoint 1.2.3.4:8000
* down-1 ---allowed-----> up-2, which has endpoint 1.2.3.4:9000

Thus, suppose we invoke this method `ListAllowedEndpointsForService(down-1, up-1)`, we expect it to return an empty list. In the current implementation [here](https://github.com/openservicemesh/osm/blob/main/pkg/catalog/endpoint.go#L64-L66), in the 3-level nested loop:

* `destSvcAccounts` => [up-2]
* `outboundEndpoints` => 1.2.3.4:8000 (endpoint of up-1)
* `podEndpoints` => [1.2.3.4:9000] (endpoints of up-2)

Because their IP matches, the `allowedEndpoints` => [1.2.3.4:8000]. This makes up-1 allowed. So in this example, even the current implementation is wrong. I think the current implementation has an assumption that different services will not have pods with the same IP, thus we only compared IP addresses.

Comment (shashankram) [reply pkg/catalog/endpoint.go:52]: Your understanding is correct. Why not simply use a mapset for storing the endpoints and doing `podEndpoints.Intersect(allowedEndpoints)`?

File: pkg/catalog/endpoint.go (hunk 1 above)
Comment (allenlsy) [reply pkg/catalog/endpoint.go]: I didn't go with the mapset approach. The key of mapset has to be hashable, while right now endpoint.Endpoint is not.

**Review by shashankram — APPROVED** [review]
Could you also add tests to cover scenarios where there are duplicate IP/port combinations for the endpoints, and simulate the bug that existed previously which this change addresses.

File: pkg/catalog/endpoint_test.go [hunk 3]
```diff
@@ -129,6 +129,28 @@ func TestListAllowedEndpointsForService(t *testing.T) {
 				Port: endpoint.Port(tests.ServicePort),
 			}},
 		},
+		{
+			name: `Traffic target defined for bookstore ServiceAccount.
+			This service account has bookstore-v1 bookstore-v2 services,
+			Bookstore-v2 pod has service account bookstore-v2 which is allowed in the traffic target, while bookstore-v1 is denied.
+			Hence, when asked about endpoints of upstream service bookstore-v1, it returns an empty list. This holds even if endpoint of v1 and v2 has the same IP address.`,
```
Comment (shashankram) [inline pkg/catalog/endpoint_test.go:136]: Use a shorter test name, and add the rest of the description as a comment.

Comment (allenlsy) [reply pkg/catalog/endpoint_test.go]: Will make it a bit shorter, but not as short as other one line names. This follows the style from other cases in this test.

File: pkg/catalog/endpoint.go [hunk 4]
```diff
@@ -38,33 +38,40 @@ func (mc *MeshCatalog) GetResolvableServiceEndpoints(svc service.MeshService) ([
...
-	destSvcAccounts, err := mc.ListAllowedOutboundServiceIdentities(downstreamIdentity)
+	destSvcIdentities, err := mc.ListAllowedOutboundServiceIdentities(downstreamIdentity)
 	if err != nil {
 		log.Error().Err(err).Msgf("Error looking up outbound service accounts for downstream identity %s", downstreamIdentity)
 		return nil, err
 	}
 
 	// allowedEndpoints comprises of only those endpoints from outboundEndpoints that matches the endpoints from listEndpointsForServiceIdentity
 	// i.e. only those interseting endpoints are taken into cosideration
+	destPodEndpointsSet := make(map[string]endpoint.Endpoint)
```
Comment (eduser25) [inline pkg/catalog/endpoint.go:63]: you are not using `destPodEndpointsSet` to really lookup anything (you end up walking it), so this map could just remain as a slice and avoid the N walk?

File: pkg/catalog/endpoint.go (hunk 2 above)
Comment (allenlsy) [reply pkg/catalog/endpoint.go]: Turns out we cannot use port number as key in the hash map to match endpoints. `ListEndpointsForIdentity()` returns endpoints without port, that causes matching fail in some cases. It's detected by our CI this time.

https://github.com/openservicemesh/osm/blob/0f792a579b05de5221d81e88399a4834bd0e7c55/pkg/providers/kube/client.go#L71-L98

File: pkg/catalog/endpoint.go (hunk 4 above)
Comment (allenlsy) [reply pkg/catalog/endpoint.go]: Right. `destPodEndpointSet` is removed.

File: pkg/catalog/endpoint.go [hunk 5]
```diff
@@ -57,16 +63,20 @@ func (mc *MeshCatalog) ListAllowedEndpointsForService(downstreamIdentity identit
...
-				}
+	for _, destSvcIdentity := range destSvcIdentities {
+		log.Info().Msgf("ups svc endpoints: %v, %v", destSvcIdentity, mc.listEndpointsForServiceIdentity(destSvcIdentity))
+
+		for _, ep := range mc.listEndpointsForServiceIdentity(destSvcIdentity) {
+			epIPStr := ep.IP.String()
+			// check if endpoint IP is allowed
+			if _, ok := outboundEndpointsSet[epIPStr]; ok {
+				// add all allowed endpoints on the pod to result list
+				// TODO(allenlsy): only allow endpoint with matching port
```
Comment (shashankram) [inline pkg/catalog/endpoint.go:74]: Please open a bug to track this.

**Review by shashankram — APPROVED** [review]

File: pkg/catalog/endpoint.go (hunk 5 above)
Comment (allenlsy) [reply pkg/catalog/endpoint.go:74]: Issue #3680

**Review by eduser25 — APPROVED** [review]

**Review by snehachhabria — APPROVED** [review]
